pkg/auth/providers/activedirectory/activedirectory_client.go

```diff
@@ -6,80 +6,73 @@ import (
 	"reflect"
 	"strings"
 
-	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
-
 	ldapv3 "github.com/go-ldap/ldap/v3"
 	"github.com/pkg/errors"
 	"github.com/rancher/norman/httperror"
 	"github.com/rancher/norman/types/slice"
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
 	"github.com/rancher/rancher/pkg/auth/providers/common/ldap"
-	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
 	"github.com/sirupsen/logrus"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 var defaultUserAttributes = []string{MemberOfAttribute, ObjectClass, ObjectGUIDAttribute}
 
-func (p *adProvider) loginUser(adCredential *v32.BasicLogin, config *v32.ActiveDirectoryConfig, caPool *x509.CertPool, testServiceAccountBind bool) (v3.Principal, []v3.Principal, error) {
+func (p *adProvider) loginUser(lConn ldapv3.Client, credentials *v3.BasicLogin, config *v3.ActiveDirectoryConfig) (v3.Principal, []v3.Principal, error) {
 	logrus.Debug("Now generating Ldap token")
 
-	username := adCredential.Username
-	password := adCredential.Password
+	password := credentials.Password
 	if password == "" {
 		return v3.Principal{}, nil, httperror.NewAPIError(httperror.MissingRequired, "password not provided")
 	}
-	externalID := ldap.GetUserExternalID(username, config.DefaultLoginDomain)
 
-	lConn, err := p.ldapConnection(config, caPool)
-	if err != nil {
-		return v3.Principal{}, nil, err
-	}
-	defer lConn.Close()
-
-	serviceAccountPassword := config.ServiceAccountPassword
-	serviceAccountUserName := config.ServiceAccountUsername
-	if testServiceAccountBind {
-		err = ldap.AuthenticateServiceAccountUser(serviceAccountPassword, serviceAccountUserName, config.DefaultLoginDomain, lConn)
-		if err != nil {
-			return v3.Principal{}, nil, err
-		}
+	sAMAccountName := credentials.Username
+	if strings.Contains(credentials.Username, `\`) {
+		sAMAccountName = strings.SplitN(credentials.Username, `\`, 2)[1]
 	}
 
-	logrus.Debug("Binding username password")
-	err = lConn.Bind(externalID, password)
+	err := ldap.AuthenticateServiceAccountUser(config.ServiceAccountPassword, config.ServiceAccountUsername, config.DefaultLoginDomain, lConn)
 	if err != nil {
-		if ldapv3.IsErrorWithCode(err, ldapv3.LDAPResultInvalidCredentials) {
-			return v3.Principal{}, nil, httperror.WrapAPIError(err, httperror.Unauthorized, "authentication failed")
-		}
-		return v3.Principal{}, nil, httperror.WrapAPIError(err, httperror.ServerError, "server error while authenticating")
-	}
-
-	samName := username
-	if strings.Contains(username, `\`) {
-		samName = strings.SplitN(username, `\`, 2)[1]
+		return v3.Principal{}, nil, err
 	}
 
-	query := fmt.Sprintf("(%v=%v)", config.UserLoginAttribute, ldapv3.EscapeFilter(samName))
-	logrus.Debugf("LDAP Search query: {%s}", query)
+	filter := fmt.Sprintf(
+		"(&(%s=%s)%s)",
+		config.UserLoginAttribute,
+		ldapv3.EscapeFilter(sAMAccountName),
+		strings.TrimSpace(config.UserLoginFilter),
```

Comment:
```suggestion
		ldapv3.EscapeFilter(strings.TrimSpace(config.UserLoginFilter)),
```
@pmatseykanets could we consider adding the [EscapeFilter](https://pkg.go.dev/github.com/go-ldap/ldap/v3#EscapeFilter) for the UserLoginFilter occurrences too ?

Reply:
You can use `EscapeFilter` to escape filter values not the filter expression.